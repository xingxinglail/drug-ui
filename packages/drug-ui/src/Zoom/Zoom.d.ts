import * as React from 'react';
import { TransitionProps } from 'react-transition-group/Transition';
import { SimpleSpread } from '..';
interface PropsExtra {
    timeout?: number | {
        enter?: number;
        exit?: number;
    };
}
export interface ZoomProps extends SimpleSpread<TransitionProps, PropsExtra> {
    children: React.ReactElement;
}
export declare const name = "Zoom";
declare const Zoom: React.FC<ZoomProps>;
export default Zoom;
