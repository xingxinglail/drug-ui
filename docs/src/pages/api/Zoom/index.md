---
filename: /packages/drug-ui/src/Zoom/Zoom.js
---

# Zoom API

<p class="description">Zoom React 组件的 API 文档。</p>

## Import

```js
import Zoom from '@drug-ui/core/Zoom';
// or
import { Zoom } from '@drug-ui/core';
```

## Props

| Name | Type | Default | Description |
|:-----|:-----|:--------|:------------|
| <span class="prop-name">children</span> | <span class="prop-type">element</span> |  | 单个子内容元素。 |
| <span class="prop-name">in</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | 如果为```true```，组件将开始进入过渡。 |
| <span class="prop-name">timeout</span> | <span class="prop-type">number &#124; { enter?: number, exit?: number }</span> | <span class="prop-default">{ enter: 225, exit: 195 }</span> | 过渡时长。 |

`ref` 被转发到根元素.

## CSS

Style sheet name: `DuiZoom`.

## Demos

- [Transitions](/drug-ui/components/Transitions)




