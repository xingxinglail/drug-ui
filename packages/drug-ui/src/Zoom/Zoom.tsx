import * as React from 'react';
import { Transition } from 'react-transition-group';
import { TransitionProps } from 'react-transition-group/Transition';
import { SimpleSpread } from '..';

interface PropsExtra {
    timeout?: number | { enter?: number, exit?: number };
}

export interface ZoomProps extends SimpleSpread<TransitionProps, PropsExtra> {
    children: React.ReactElement;
}

export const name = 'Zoom';

const Zoom: React.FC<ZoomProps> = React.forwardRef<HTMLDivElement, ZoomProps>((props, ref) => {
    const {
        style,
        in: inProp = false,
        timeout = {
            enter: 225,
            exit: 195
        },
        onEnter,
        onEntering,
        onEntered,
        onExit,
        onExiting,
        onExited,
        children,
        ...rest
    } = props;

    const handleTransitionEvent = (event: string, node: HTMLElement, isAppearing?: boolean) => {
        if (event === 'onEnter') {
            node.style.transform = 'scale(0)';
            onEnter && onEnter(node, isAppearing);
        }

        if (event === 'onEntering') {
            node.offsetTop;
            node.style.transform = 'scale(1)';
            node.style.transition = `transform ${ typeof timeout === 'number' ? timeout : timeout.enter }ms cubic-bezier(0.4, 0, 0.2, 1)`;
            onEntering && onEntering(node, isAppearing);
        }

        if (event === 'onEntered') {
            onEntered && onEntered(node, isAppearing);
        }

        if (event === 'onExit') {
            node.style.transform = 'scale(1)';
            onExit && onExit(node);
        }

        if (event === 'onExiting') {
            node.style.transform = 'scale(0)';
            node.style.transition = `transform ${ typeof timeout === 'number' ? timeout : timeout.exit }ms cubic-bezier(0.4, 0, 0.2, 1)`;
            onExiting && onExiting(node);
        }

        if (event === 'onExited') {
            onExited && onExited(node);
        }
    };

    return (
        <Transition
            in={ inProp }
            timeout={ timeout }
            onEnter={ (node, isAppearing) => handleTransitionEvent('onEnter', node, isAppearing) }
            onEntering={ (node, isAppearing) => handleTransitionEvent('onEntering', node, isAppearing) }
            onEntered={ (node, isAppearing) => handleTransitionEvent('onEntered', node, isAppearing) }
            onExit={ (node) => handleTransitionEvent('onExit', node) }
            onExiting={ (node) => handleTransitionEvent('onExiting', node) }
            onExited={ (node) => handleTransitionEvent('onExited', node) }
            { ...rest }>
            {
                (state: string, childProps: any) => {
                    return React.cloneElement(children, {
                        style: {
                            visibility: state === 'exited' && !inProp ? 'hidden' : undefined,
                            ...style,
                            ...children.props.style
                        },
                        ...childProps,
                        ref
                    })
                }
            }
        </Transition>
    );
});

Zoom.displayName = name;

export default Zoom;
