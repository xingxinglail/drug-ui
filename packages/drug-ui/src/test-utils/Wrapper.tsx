import * as React from 'react';
import { ThemeProvider } from '../index';

interface Props {
    children?: any;
}

const Wrapper: React.FC<Props> = (props) => {
    return (
        <ThemeProvider>
            { props.children }
        </ThemeProvider>
    );
};

export default Wrapper;
