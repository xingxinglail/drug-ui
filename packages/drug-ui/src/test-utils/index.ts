export { default as Wrapper } from './Wrapper';
export { getClasses } from './getClasses';

export const wait = (delay = 50) => {
    return new Promise(resolve => setTimeout(resolve, delay));
};
