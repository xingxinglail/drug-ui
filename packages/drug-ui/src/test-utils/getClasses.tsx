import * as React from 'react';
import * as enzyme from 'enzyme';
import Wrapper from './Wrapper';

export const getClasses = (element: React.ReactElement) => {
    // @ts-ignore
    const { useStyles } = element.type;
    let classes;

    function Listener () {
        classes = useStyles();
        return null;
    }

    enzyme.mount(<Wrapper><Listener /></Wrapper>);
    return classes;
};
