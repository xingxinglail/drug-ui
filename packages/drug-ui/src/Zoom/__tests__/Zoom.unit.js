import * as renderer from 'react-test-renderer';
import * as React from 'react';
import * as enzyme from 'enzyme';
import Zoom from '..';
import { Wrapper, wait } from '@drug-ui/core/test-utils';

const mount = enzyme.mount;

describe('<Zoom />', () => {

    it('render successful', () => {
        const json = renderer.create(<Wrapper><Zoom><div /></Zoom></Wrapper>).toJSON();
        expect(json).toMatchSnapshot();
    });

    it('默认渲染', () => {
        const wrapper = mount(<Wrapper><Zoom><div /></Zoom></Wrapper>);
        expect(window.getComputedStyle(wrapper.getDOMNode()).visibility).toEqual('hidden');
    });

    it('默认显示渲染', () => {
        const wrapper = mount(<Wrapper><Zoom in><div /></Zoom></Wrapper>);
        expect(window.getComputedStyle(wrapper.getDOMNode()).visibility).toEqual('');
    });

    it('触发所有钩子', (done) => {
        const onEnter = jest.fn();
        const onEntering = jest.fn();
        const onEntered = jest.fn(async () => {
            expect(window.getComputedStyle(dom).getPropertyValue('visibility')).toEqual('');
            expect(window.getComputedStyle(dom).getPropertyValue('transform')).toEqual('scale(1)');
            await wait();
            wrapper.setProps({ in: false });
        });
        const onExit = jest.fn();
        const onExiting = jest.fn();
        const onExited = jest.fn(() => {
            expect(window.getComputedStyle(dom).getPropertyValue('visibility')).toEqual('hidden');
            expect(window.getComputedStyle(dom).getPropertyValue('transform')).toEqual('scale(0)');
            expect(onEnter).toHaveBeenCalledBefore(onEntering);
            expect(onEntering).toHaveBeenCalledBefore(onEntered);
            expect(onEntered).toHaveBeenCalledBefore(onExit);
            expect(onExit).toHaveBeenCalledBefore(onExiting);
            expect(onExiting).toHaveBeenCalledBefore(onExited);
            done();
        });

        const wrapper = mount(
            <Wrapper>
                <Zoom
                    onEnter={ onEnter }
                    onEntering={ onEntering }
                    onEntered={ onEntered }
                    onExit={ onExit }
                    onExiting={ onExiting }
                    onExited={ onExited }>
                    <div />
                </Zoom>
            </Wrapper>
        );
        const dom = wrapper.getDOMNode();
        expect(window.getComputedStyle(dom).getPropertyValue('visibility')).toEqual('hidden');
        wrapper.setProps({ in: true });
    });
});
