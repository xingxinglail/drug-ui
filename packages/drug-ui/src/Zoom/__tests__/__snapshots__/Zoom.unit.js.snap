// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Zoom /> render successful 1`] = `
<div
  style={
    Object {
      "visibility": "hidden",
    }
  }
/>
`;
